### PR TITLE
Bootstrap From Snapshot Pt. 3: Test Coverage

### PR DESCRIPTION
**Pull request by jordanschalm**
Ticket: https://github.com/dapperlabs/flow-go/issues/4890

This is part 3 of supporting bootstrapping from an arbitrary protocol state snapshot (following from #272 and https://github.com/onflow/flow-go/pull/312).

#### Changes

* Add tests for bootstrapping the protocol state from non-root snapshots
* Change `Snapshot` interface to combine seal/result query to `SealedResult`  https://github.com/onflow/flow-go/pull/312#discussion_r571633767
* Split `Bootstrap` function into helper functions
* Fix a bug in `inmem.Snapshot#Clustering` where it would not return clustering for set-up but un-committed epochs
* Update various test utilities and tests that made assumptions broken by bootstrapping changes
* Change `PayloadFixture`:
  * Return literal rather than pointer to match block/header fixtures and so it can be used directly in `SetPayload`
  * Use an empty payload by default and add contents as needed. An empty payload works fine for most applications. For cases where we needed the payload to have a particular to have a particular form, it has been updated to explicitly specify that when generating the fixture (eg. `PayloadFixture(WithSeals(...))`).

### CODE REVIEW

**Review by zhangchiqing — APPROVED** [review]
Looks good 👍

**Review by AlexHentschel — APPROVED** [review]
nice

File: state/protocol/badger/state_test.go [hunk 1]
```diff
@@ -43,12 +46,289 @@ func TestBootstrapAndOpen(t *testing.T) {
...
+				require.NoError(t, err)
+				if phase == flow.EpochPhaseSetup && counter == epoch1Counter+1 {
+					return snap
+				}
+			}
+		})
+
+		bootstrap(t, rootSnapshot, func(state *bprotocol.State, err error) {
+			require.NoError(t, err)
+			assert.NoError(t, err)
```
Comment (jordanschalm) [inline state/protocol/badger/state_test.go:171]:
```suggestion
			require.NoError(t, err)
```

[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/state_test.go]: Done in https://github.com/onflow/flow-go/pull/374/commits/301d10b72f6dca63c1d9c068a097e5961c3ce015 (GH wouldn't let me apply this suggestion on the site for some reason 🤔 )

**Pull request merged**